Use a lua expression as the tabline variable (and other enhancements)

jdrouhard opened the pull request:
Use a lua expression as the tabline variable. This means we don't
have to rely on autocmd events to do a rebuild of the tabline since we
can just rely on nvim's built-in re-eval logic.

This both improves performance as well as correctness. The tabline will
only be redrawn when neovim knows something has changed. This includes
saving buffers (so they are no longer modified) on non-visible buffers
(fixes #21).

Because we aren't repeatedly updating the tabline option (relying on the
built-in eval logic), the start/help screen when vim is first opened is
no longer immediately erased which fixes #22.

Loading buftabline will now occur on VimEnter if it is autoloaded at
startup, but will occur synchronously if re-loaded or if it's
lazy-initialized.

jdrouhard commented on lua/buftabline/commands.lua:
```diff
@@ -56,22 +56,11 @@ M.on_buffer_delete = b.on_buffer_delete
-        u.clear_augroup()
-    end)
-end
+M.build = require("buftabline.build")
```
Probably don't need this anymore, and can just use `require("buftabline.build")()` directly in the global function defined in init.lua. I can make this so if you wish.

jose-elias-alvarez replied:
Agreed that there's no need for this and we can delete it.

jose-elias-alvarez commented:
This is actually how the plugin worked until recently, and while it does improve correctness, what I found is that Neovim is incredibly aggressive about when it calls the expression, to the point that the editor would occasionally crash when doing something that triggered a lot of calls (e.g. scrolling through Telescope previews). Are you sure that this has a positive effect on performance? 

The first linked issue will be solved when the PR is merged (and we can use BufWritePost in the meantime), and I'm not sure that the second issue is serious enough to warrant this change, so I'd have to be sure that this won't negatively affect performance (which goes against my prior experience and is why I changed it in the first place).

jdrouhard commented:
I'm fairly certain this improves it, though I can't say for certain. Changing the tabline options itself repeatedly seems to slow it down more than the aggressive eval. It definitely feels better to me after this change. Are you sure that Neovim's built-in logic for determining when to re-eval is too aggressive? If it is, the actual build() function can be optimized so it doesn't have to do as much work by keeping a complete cache of buffers/tabs on autocmd events (but not actually redrawing). This is already partially complete with the state stuff you added recently--it could be augmented to store full information so the build() function can just assemble the string and nothing else.

Every other tabline plugin I looked at it does it this way and they don't have any performance issues (that I'm aware of). vim-airline is especially snappy and I know it builds a cache based on autocmd events and then builds the string on the eval function.

jdrouhard commented:
Hmm, GitHub experienced an issue when I was trying to update this PR. I fixed the tests and changed a couple other things but it's not reflected here.

jose-elias-alvarez commented on lua/buftabline/init.lua:
```diff
@@ -3,33 +3,54 @@ local u = require("buftabline.utils")
 
 local M = {}
 
+function _G.buftabline()
```
Can we use `luaeval` to avoid having to define a global variable? This is how we did it before:

```lua
vim.o.tabline = [[%!luaeval('require("buftabline").build()')]]
```

jdrouhard replied:
Works for me. I'm still learning the ropes of nvim's lua integration. It's pretty incredible but there's lots there!

jose-elias-alvarez commented:
It's hard to quantify, but it does seem to feel a little nicer with this change, and honestly, I like removing the overhead of deciding when to update the tabline, even if it might result in unnecessary re-renders .I think using `vim.schedule` (another change made recently) should take care of crashes, and I'll see what I can do about changing how `build` works, too, since I think that's a logical change.

Let me know what you think about using `luaeval` to avoid the global variable. Other than that, I'm happy to merge this.

jdrouhard commented:
Yeah, I was thinking more about why it felt snappier, and I think it's due to how the current code delays updating the tabline text by a full event loop iteration. Here's my guess as to what's happening--neovim's internal event loop processes something (like a BufAdd), then decides if it needs to re-evaluate status and tablines, does so if necessary, and _then loops through and fires the autocmd events_. The autocmd callbacks would hit the code that simply schedules another callback on neovim's event loop, so you're at least at the mercy of waiting until all the autocmd callbacks have finished. It likely takes a full loop iteration with everything involved there before it even gets around to actually building the tabline and updating it. There might even be some overhead to setting a new tabline text via the `tabline` option but I'm not certain on that. Depending on where in the loop custom callbacks are processed, the redraw might be soon after the autocmds themselves or it might not be for nearly a full loop iteration later.

This is all conjecture without looking at neovim's source, but seems to support the evidence.

jose-elias-alvarez commented:
No idea, but that sounds like a reasonable explanation. I just pushed a couple of changes to fix tests and also fix `check_auto_hide` (scheduling it is an easy way to make sure it's accurate, since the event timing is weird).

Thanks for working on this!

jdrouhard commented:
Thanks for fixing the tests after I removed the global variable! scheduling check_auto_hide() seems like a good change too.

jose-elias-alvarez commented:
So while working today, I found issues with two pieces of functionality that I use a lot (auto hide and tabpage buffers). I imagine it has to do with the timing of autocommand events, but one of the advantages of scheduling updates was to avoid worrying about that. I've reverted the changes (my fault for not testing them more thoroughly) and moved the changes in this PR to [this branch](https://github.com/jose-elias-alvarez/buftabline.nvim/tree/expr-tabline).

If you don't care about those features and are happy with the plugin as-is, I recommend forking it (I don't think it'll change too much either way). If we want to go back to the Lua expression method, we'd need to investigate and fix the issues and test the functionality more comprehensively. The current suite asserts checks whether the tablline is updated correctly after each relevant event, so at a minimum I'd want to replicate that.

jdrouhard commented:
I'd like to see these changes stay in, so I can take a look at fixing the issues you ran into. I could maintain a fork, but I'd rather contribute to the main repository in the interest of open source :)

jose-elias-alvarez commented:
I appreciate your interest. I want to mention that after recent updates, I mentally considered this plugin "done" and didn't really plan on doing anything with it (other than bug fixes) so I could focus on other projects. Since it's working as-is, my priority is to avoid breaking it, and so I don't know if I have much of an appetite for another round of changes. 

The specific issues I saw seemed related to the autocommands in `buffers.lua`, since what I experience was that the tabline wasn't correctly updated when buffers were added / deleted with multiple tabpages open. Again, if we're able to thoroughly test that these events correctly update, I'm not against merging it, but as I said, my priority is to avoid breaking it so I can rely on it and work on (frankly more interesting) projects.

jdrouhard commented:
Understood. I do appreciate you writing a really good tabline plugin. It's basically the exact feature set I was looking for (the others are too heavy or didn't quite have what I wanted).

If you are happy with the way it is, then I understand your desire to not change things for the sake of some other random user on the internet. :) I can maintain my fork, I'll just need to remember to check in occasionally for other tune ups or features you add. Typically I like to avoid maintaining separate forks since I adhere to the OSS philosophy of sharing improvements.

My guess on what the issue is that there are probably a couple missing 'redrawtabline' commands in some autocmd handlers. I did notice that vim-airline has a `redrawtabline` call in the `BufAdd` autocmd handler but didn't look too closely as to why. I bet just adding `redrawtabline` to the end of `on_buf_add()` would fix your issues, but we'd need to reproduce in a test and prove it works, which is hard to do when the tests can only see what `build()` would produce (currently).

jose-elias-alvarez commented:
Thanks for understanding!
